x/programs/cmd/simulator/src/simulator.rs

```diff
@@ -0,0 +1,238 @@
+        .spawn()
+        .expect("unable to spawn simulator");
+
+    let writer = stdin.expect("unable to get stdin");
+    let reader = stdout.expect("unable to get stdout");
+
+    let responses = BufReader::new(reader)
+        .lines()
+        .map(|line| serde_json::from_str(&line?).map_err(SimulatorError::Serde));
+
+    Simulator { writer, responses }
+}
+
+impl<W, R> Simulator<W, R>
+where
+    W: Write,
+    R: Iterator<Item = SimulatorResponseItem>,
+{
+    const RUN_COMMAND: &'static [u8] = b"run --message '";
+
+    pub fn create_program<P: AsRef<Path>>(&mut self, path: P) -> SimulatorResponseItem {
+        self.writer.write_all(Self::RUN_COMMAND)?;
+
+        let path = path.as_ref().to_string_lossy();
+        let input = serde_json::to_vec(&SimulatorRequest::create_program(path.into()))?;
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+    }
+
+    pub fn read(&mut self, method: String, params: Vec<Param>) -> SimulatorResponseItem {
+        self.writer.write_all(Self::RUN_COMMAND)?;
+
+        let input = serde_json::to_vec(&SimulatorRequest::read(method, params))?;
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+    }
+
+    pub fn execute(
+        &mut self,
+        method: String,
+        params: Vec<Param>,
+        max_units: u64,
+    ) -> SimulatorResponseItem {
+        self.writer.write_all(Self::RUN_COMMAND)?;
+
+        let input = serde_json::to_vec(&SimulatorRequest::execute(method, params, max_units))?;
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+    }
+}
+
+/// A [`SimulatorRequest`] is a call to the simulator
+#[derive(Debug, Serialize, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+struct SimulatorRequest {
+    /// The API endpoint to call.
+    endpoint: Endpoint,
+    /// The method to call on the endpoint.
+    method: String,
+    /// The maximum number of units the Request can consume.
+    max_units: u64,
+    /// The parameters to pass to the method.
+    params: Vec<Param>,
+}
+
+impl SimulatorRequest {
+    pub fn read(method: String, params: Vec<Param>) -> Self {
+        Self {
+            endpoint: Endpoint::ReadOnly,
+            method: method.to_string(),
+            max_units: 0,
+            params,
+        }
+    }
+
+    pub fn execute(method: String, params: Vec<Param>, max_units: u64) -> Self {
+        Self {
+            endpoint: Endpoint::Execute,
+            method,
+            max_units,
+            params,
+        }
+    }
```

Comment:
Not sure why we used to pass in a String, a &str would work just fine !

```suggestion
impl<'a> SimulatorRequest<'a> {
    pub fn read(method: &'a str, params: Vec<Param>) -> Self {
        Self {
            endpoint: Endpoint::ReadOnly,
            method,
            max_units: 0,
            params,
        }
    }

    pub fn execute(method: &'a str, params: Vec<Param>, max_units: u64) -> Self {
        Self {
            endpoint: Endpoint::Execute,
            method,
            max_units,
            params,
        }
    }
```

Reply:
Then you could change the signature of `Simulator::read` and `Simulator::write` so that they take the a lifetime as `Simulator::read<'a>` and `Simulator::write<'a>` and pass it to the `method` parameter. This suggestion seems a bit out of scope of this PR though so that could be done as a separate one.

Reply:
will do in a follow up, made an issue
https://github.com/ava-labs/hypersdk/issues/1169